File: day_11/src/com/brahle/adventofcode/year2019/IntMachine.scala
package com.brahle.adventofcode.year2019
import scala.collection.mutable

class IntMachine {
  private var position = BigInt.int2bigInt(0);
  private var relativeIndex = BigInt.int2bigInt(0);
  private var memory = mutable.Map[BigInt, BigInt]();
  private var inputs = mutable.Queue[BigInt]()
  var input: () => BigInt = () => { inputs.dequeue() };
  var output: (BigInt) => Unit = (x: BigInt) => { println("OUTPUT: " + x.toString()) };

  def this(values: String, inputs: String)
  {
    this();
    values
      .split(",")
      .zipWithIndex
      .foreach(tuple => {
        val pos = BigInt.int2bigInt(tuple._2);
        val value = BigInt(tuple._1);
        memory(pos) = value;
      });
    if (inputs != "") {
      inputs
        .split(",")
        .foreach(value => this.inputs.enqueue(BigInt(value)));
    }
  }

  def evaluate(): Unit = {
    while (evaluateNext()) {}
  }

  def evaluateNext(): Boolean = {
    operation() match {
      case 1 => op1()
      case 2 => op2()
      case 3 => op3()
      case 4 => op4()
      case 5 => op5()
      case 6 => op6()
      case 7 => op7()
      case 8 => op8()
      case 9 => op9()
      case 99 => op99()
    }
  }

  def op1(): Boolean = {
    write(3, resolve(1) + resolve(2))
    position += 4
    true
  }

  def op2(): Boolean = {
    write(3, resolve(1) * resolve(2))
    position += 4
    true
  }

  def op3(): Boolean = {
    write(1, input())
    position += 2
    true
  }

  def op4(): Boolean = {
    output(resolve(1))
    position += 2
    true
  }

  def op5(): Boolean = {
    if (resolve(1) != 0) {
      position = resolve(2)
    } else {
      position += 3
    }
    true
  }

  def op6(): Boolean = {
    if (resolve(1) == 0) {
      position = resolve(2)
    } else {
      position += 3
    }
    true
  }

  def op7(): Boolean = {
    if (resolve(1) < resolve(2)) {
      write(3, 1)
    } else {
      write(3, 0)
    }
    position += 4
    true
  }

  def op8(): Boolean = {
    if (resolve(1) == resolve(2)) {
      write(3, 1)
    } else {
      write(3, 0)
    }
    position += 4
    true
  }

  def op9(): Boolean = {
    relativeIndex += resolve(1)
    position += 2
    true
  }

  def op99(): Boolean = {
    println("DONE " + memory(0).toString())
    false
  }

  def op(): Int = memory(position).toInt

  def operation(): Int = op() % 100

  def resolve(index: Int): BigInt = {
    mode(index) match {
      case 0 => positional(index)
      case 1 => immediate(index)
      case 2 => relative(index)
    }
  }

  def positional(index: Int): BigInt = {
    read(memory(position + index))
  }

  def immediate(index: Int): BigInt = {
    read(position + index)
  }

  def relative(index: Int): BigInt = {
   read(relativeIndex + memory(position + index))
  }

  def read(x: BigInt): BigInt = {
    memory getOrElse(x, 0)
  }

  def write(index: Int, value: BigInt) = {
    memory(getWriteIndex(index)) = value
  }

  def getWriteIndex(index: Int): BigInt = {
    mode(index) match {
      case 0 => read(position + index)
      case 2 => read(position + index) + relativeIndex
    }
  }

  def mode(index: Int): Int = {
    index match {
      case 1 => op() / 100 % 10
      case 2 => op() / 1000 % 10
      case 3 => op() / 10000 % 10
      case 4 => op() / 100000 % 10
    }
  }
}

File: day_15/intmachine.rs
use std::collections::HashMap;
use std::vec::Vec;

pub struct IntMachine<F1, F2> where
        F1: FnMut() -> i64,
        F2: FnMut(i64) -> () {
    pub position: i64,
    pub relative_index: i64,
    pub memory: HashMap<i64, i64>,
    pub input: F1,
    pub output: F2,
}

pub fn sample_input() -> i64 {
    0
}

pub fn sample_output(x: i64) {
    println!("OUTPUT: {}", x);
}

pub fn machine(
    memory: HashMap<i64, i64>,
    input: impl FnMut() -> i64,
    output: impl FnMut(i64) -> (),
) -> IntMachine<impl FnMut() -> i64, impl FnMut(i64) -> ()> {
    IntMachine {
        position: 0,
        relative_index: 0,
        memory: memory,
        input: input,
        output: output,
    }
}

pub fn queue_input(q: Vec<i64>) -> impl FnMut() -> i64 {
    let mut i = 0;
    let _input = move || {
        i += 1;
        return q[i-1];
    };
    return _input;
}


impl<F1, F2> IntMachine<F1, F2> where
        F1: FnMut() -> i64,
        F2: FnMut(i64) -> () {

    pub fn evaluate(&mut self) {
        while self.evaluate_next() {};
    }

    fn evaluate_next(&mut self) -> bool {
        match self.operation() {
            1 => return self.op1(),
            2 => return self.op2(),
            3 => return self.op3(),
            4 => return self.op4(),
            5 => return self.op5(),
            6 => return self.op6(),
            7 => return self.op7(),
            8 => return self.op8(),
            9 => return self.op9(),
            _ => return self.op99(),
        }
    }

    fn op1(&mut self) -> bool {
        // self.write(&3, self.resolve(1) + self.resolve(2));
        self.memory.insert(self.write_index(&3), self.resolve(1) + self.resolve(2));
        self.position += 4;
        return true;
    }

    fn op2(&mut self) -> bool {
        self.memory.insert(self.write_index(&3), self.resolve(1) * self.resolve(2));
        self.position += 4;
        return true;
    }

    fn op3(&mut self) -> bool {
        self.memory.insert(self.write_index(&1), (self.input)());
        self.position += 2;
        return true;
    }

    fn op4(&mut self) -> bool {
        let x = *self.resolve(1);
        (self.output)(x);
        self.position += 2;
        return true;
    }

    fn op5(&mut self) -> bool {
        if *self.resolve(1) != 0 {
            self.position = *self.resolve(2);
        } else {
            self.position += 3;
        }
        return true;
    }

    fn op6(&mut self) -> bool {
        if *self.resolve(1) == 0 {
            self.position = *self.resolve(2);
        } else {
            self.position += 3;
        }
        return true;
    }

    fn op7(&mut self) -> bool {
        if *self.resolve(1) < *self.resolve(2) {
            self.memory.insert(self.write_index(&3), 1);
        } else {
            self.memory.insert(self.write_index(&3), 0);
        }
        self.position += 4;
        return true;
    }

    fn op8(&mut self) -> bool {
        if *self.resolve(1) == *self.resolve(2) {
            self.memory.insert(self.write_index(&3), 1);
        } else {
            self.memory.insert(self.write_index(&3), 0);
        }
        self.position += 4;
        return true;
    }

    fn op9(&mut self) -> bool {
        self.relative_index += *self.resolve(1);
        self.position += 2;
        return true;
    }

    fn op99(&self) -> bool {
        println!("DONE {}", self.memory.get(&0).unwrap());
        return false;
    }

    fn op(&self) -> &i64 {
        self.memory.get(&self.position).unwrap()
    }

    fn operation(&self) -> i64 {
        self.op() % 100
    }

    fn resolve(&self, index: i64) -> &i64 {
        match self.mode(index) {
            0 => return self.positional(index),
            1 => return self.immediate(index),
            _ => return self.relative(index),
        }
    }

    fn mode(&self, index: i64) -> i64 {
        match index {
            1 => return self.op() / 100 % 10,
            2 => return self.op() / 1000 % 10,
            3 => return self.op() / 10000 % 10,
            _ => return self.op() / 100000 % 10,
        }
    }

    fn positional(&self, index: i64) -> &i64 {
        return self.read(self.memory.get(&(self.position + index)).unwrap());
    }

    fn immediate(&self, index: i64) -> &i64 {
        return self.read(&(self.position + index));
    }

    fn relative(&self, index: i64) -> &i64 {
        return self.read(&(
            self.relative_index +
            self.memory.get(&(self.position + index)).unwrap()
        ));
    }

    fn read(&self, index: &i64) -> &i64 {
        return self.memory.get(index).unwrap_or(&0);
    }

    fn write(mut self, index: &i64, value: i64) {
        self.memory.insert(self.write_index(index), value);
    }

    fn write_index(&self, index: &i64) -> i64 {
        match self.mode(*index) {
            0 => return *self.read(&(self.position + index)),
            _ => return self.read(&(self.position + index)) + self.relative_index,
        }
    }
}
